internal/dot/graph.go

```diff
@@ -252,6 +274,59 @@ func (dg *Graph) addToGroup(r *Result, id CtorID) {
 	group.Results = append(group.Results, r)
 }
 
+// Prune removes elements from the graph that do not have failed results.
+// Removing elements that do not have failing results makes the graph easier to debug, since non-failing
+// nodes and edges can clutter the graph and don't help the user debug.
+func (dg *Graph) Prune() {
+	dg.pruneCtors()
+	dg.pruneGroups()
+}
+
+// Remove constructors from the graph that do not have failing Results.
+func (dg *Graph) pruneCtors() {
+	var prunedCtors []*Ctor
+	for _, c := range dg.Ctors {
+		if _, ok := dg.Failed.FailedCtors[c.ID]; ok {
+			prunedCtors = append(prunedCtors, c)
+		} else {
+			dg.pruneGroupResults(c)
+			delete(dg.ctorMap, c.ID)
+		}
+	}
+
+	dg.Ctors = prunedCtors
+}
+
+// Remove groups from the graph that do not have failing Results.
+func (dg *Graph) pruneGroups() {
+	var prunedGroups []*Group
+	for _, g := range dg.Groups {
+		if _, ok := dg.Failed.FailedGroups[groupKey{t: g.Type, group: g.Name}]; ok {
+			prunedGroups = append(prunedGroups, g)
+		} else {
+			delete(dg.groupMap, groupKey{t: g.Type, group: g.Name})
+		}
+	}
+
+	dg.Groups = prunedGroups
+}
+
+// PruneGroupResults removes constructor results that are still referenced in
+// their respective Group object.
+func (dg *Graph) pruneGroupResults(c *Ctor) {
+	for _, r := range c.Results {
+		var prunedResults []*Result
+		if resultGroup, ok := dg.groupMap[r.groupKey()]; ok {
+			for _, gr := range resultGroup.Results {
+				if r.GroupIndex != gr.GroupIndex {
+					prunedResults = append(prunedResults, gr)
+				}
```

Comment:
So previously, GroupIndex existed solely to help GraphViz. With this change,
it transitions into an almost-unique identifier for the Result.

In a follow-up PR, consider refactoring this a bit so that we have a ResultID
type (which is probably just an integer).

```
type ResultID int
```

The uniqueness of the ResultID can be scoped to the Graph so you would track
the "next" group ID on the Graph object rather than based on the length of the
Group.Results.

```
type Graph struct {
    ...
    
    resultIDCounter int
}

func (g *Graph) nextResultID() int {
    c := g.resultIDCounter
    g.resultIDCounter++
    return c
}
```

This will let you change the Group itself to hold a `map[ResultID]*Result`,
which makes this particular deletion a lot more efficient.

---

To clarify, this is fine as-is for now. The refactoring should be done in a
future change. For now, you can just add a TODO to do this more efficiently.

```
// TODO: Avoid linear search
```

Reply:
Will add these changes to the grafx: The next generation ticket!